json: json_elements: add generate_array

bhalevy opened the pull request:
Add a helper function to generate a json array gently by consuming its items from a coroutine generator.

The motivation for this change is to avoid large allocations that are needed for generating a long json array (See https://github.com/scylladb/scylladb/pull/24261).

With the coroutine generator, we can consume and process the array elements one at a time (optionally using a small buffer to reduce context switches between the generator and the consumer), and stream them as json, say to an output file, or to a socket, as part of a higher level protocol.

bhalevy commented on include/seastar/json/json_elements.hh:
```diff
@@ -367,6 +370,45 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
+                if (first) {
+                    first = false;
+                } else {
+                    co_await out.write(", ");
```
eh?

bhalevy replied:
If it's a long array we would like to prevent both large allocations and reactor stalls, hence the generator.

robertbindar reviewed (approved):
Nice, I know understand what you wanted in https://github.com/scylladb/scylladb/pull/24261/
Please see the comment if it's possible to accommodate that.

bhalevy commented on include/seastar/json/json_elements.hh:
```diff
@@ -367,6 +370,45 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
+ */
+template<typename T, typename Func>
+requires requires (T val, Func aa, output_stream<char> s) { { formatter::write(s, aa(val)) } -> std::same_as<future<>>; }
+std::function<future<>(output_stream<char>&&)> generate_array(coroutine::experimental::generator<T>& gen, Func fun) {
```
> Can we instead make the return function get a reference to the output stream and not make it responsible for closing the stream? I'll use this for streaming an object like: `{ "key": [el1, el2, el3]}` That means I'll need to open a stream, write "{", use the output function of `generate_array` to get `[...]` and then open stream once again to write "}" and close.

I'm not sure what you mean.
The return function does get a (rvalue) reference to the output stream:
`std::function<future<>(output_stream<char>&&)>`

bhalevy replied:
Hmm, but we want to also enclose the json data in `{ }`...

bhalevy replied:
The problem is with the expectation of returning a `json_return_type` with the returned function a its `_body_writer`

bhalevy replied:
All that said, I see that `write_to_stream_and_close` closes the output stream too after invoking the writer.
@xemul can you please explain?

robertbindar replied:
> The problem is with the expectation of returning a `json_return_type` with the returned function a its `_body_writer`

we can template out `json_return_type` with default on `output_stream<char>&&` so no outside code needs changing. It depends how far you want to go to stay consistent with the other functions. These functions seems more like utility functions rather than an API, and for `generate_array` the only user is in https://github.com/scylladb/scylladb/pull/24261/, it's probably alright if we diverge from the pattern of the other functions.

robertbindar replied:
> I'll send a refactored version that uses a utility function like generate_array_element that will encode the array onto the output stream without consuming and closing the syream itself so it can be used as a building block.

looking forward to it

bhalevy replied:
See v2 ([1e79993](https://github.com/scylladb/seastar/commit/1e79993415a7a9203fb0c80f7634ad4ac142060c))

avikivity commented on include/seastar/json/json_elements.hh:
```diff
@@ -367,6 +370,47 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
+ */
+template<typename T, typename Func>
+requires requires (T val, Func aa, output_stream<char> s) { { formatter::write(s, aa(val)) } -> std::same_as<future<>>; }
+future<> generate_array_element(output_stream<char>& out, coroutine::experimental::generator<T>& gen, Func fun) {
```
Looking at std::generator, it implements std::ranges::view_interface, so you normally wouldn't consume a generator, you consume a view. The user can pass anything that's a view (like std::views::iota), or a generator if they like.

We should have the concept of a streaming view: something that has std::optional\<T> operator()().

robertbindar replied:
> Why generate_array_element? It's writing the whole array, no?

I think `element` here means `json element`, which is still confusing, but I don't have a better idea (maybe generate_json_array_type?), i.e. `{ "some_array": generage_array_element() }`

bhalevy replied:
> Why generate_array_element? It's writing the whole array, no?

The name isn't great, it's not generating an element in an array, but rather the whold array as a json element.

bhalevy replied:
> Why an extra mapping? Seems like it can be applied outside.

Hmm, the interface evolved from stream_range_as_array hence the mapping function, but I agree that using a generator indeed allows the user to compose the mapper externally as part of the generator.

bhalevy replied:
> Looking at std::generator, it implements std::ranges::view_interface, so you normally wouldn't consume a generator, you consume a view. The user can pass anything that's a view (like std::views::iota), or a generator if they like.
> 
> We should have the concept of a streaming view: something that has std::optional<T> operator()().

But it's an async generator so the view should have `future<std::optional<T>> operator()()`

avikivity replied:
Yes, I forgot the future there.

avikivity replied:
The mapping is still there.

bhalevy replied:
It's simple on paper but the optional result needs to be moved too so it needs to hold a std::unique_ptr<json_base>

avikivity replied:
How can fun return json_base if the generator can't?

bhalevy replied:
> How can fun return json_base if the generator can't?

It doesn't.  It returns an object derived from json_base that isn't moved further, but rather "serialized" onto the json stream.

I think I have a solution by allowing the implementing a noexcept move constructor for the test's object_json
and defining the generator as returning object_json with the right contraints.

I'll post a new version

bhalevy replied:
Done in [1ec981a](https://github.com/scylladb/seastar/commit/1ec981a2ac274ce5478eed374c0ea40767d1b7c4)

avikivity commented on include/seastar/json/json_elements.hh:
```diff
@@ -367,6 +370,47 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
+
+/*!
+ * \brief consume values from a coroutine generator \c gen and return a serialize function for it as a json array.
+ *
```
What's a serialize function? Is it an existing concept?

avikivity replied:
But what's a serialize function?

bhalevy replied:
I agree the name is bad.
This is actually a view transformation function from the native type T to a jsonable object.

bhalevy commented:
A simpler approach could be:
```c++
diff --git a/include/seastar/json/json_elements.hh b/include/seastar/json/json_elements.hh
index ed4b6104..1dcf4719 100644
--- a/include/seastar/json/json_elements.hh
+++ b/include/seastar/json/json_elements.hh
@@ -367,6 +367,31 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
     };
 }
 
+template <class Generator>
+concept AsyncJsonGenerator = requires (Generator gen, output_stream<char> out) {
+    { gen.has_value() } -> std::same_as<future<bool>>;
+    { gen.write(out) } -> std::same_as<future<>>;
+};
+
+/*!
+ * \brief consume values from a coroutine generator \c gen and encode the items using the \c fun mapping function
+ * onto the \c out output_stream as a json array.
+ */
+template <AsyncJsonGenerator Gen>
+future<> generate_json_array(output_stream<char>& out, Gen& gen) {
+    bool first = true;
+    co_await out.write("[");
+    while (co_await gen.has_value()) {
+        if (first) {
+            first = false;
+        } else {
+            co_await out.write(", ");
+        }
+        co_await gen.write(out);
+    }
+    co_await out.write("]");
+}
+
 /*!
  * \brief capture an object and return a serialize function for it.
  *
diff --git a/tests/unit/json_formatter_test.cc b/tests/unit/json_formatter_test.cc
index 69c820e1..11e1d501 100644
--- a/tests/unit/json_formatter_test.cc
+++ b/tests/unit/json_formatter_test.cc
@@ -25,6 +25,8 @@
 #include <seastar/testing/test_case.hh>
 #include <seastar/core/sstring.hh>
 #include <seastar/core/vector-data-sink.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/coroutine/generator.hh>
 #include <seastar/json/formatter.hh>
 #include <seastar/json/json_elements.hh>
 #include <seastar/testing/thread_test_case.hh>
@@ -135,6 +137,35 @@ SEASTAR_THREAD_TEST_CASE(test_stream_range_as_array) {
     }, false);
 }
 
+SEASTAR_THREAD_TEST_CASE(test_generate_array) {
+    sstring expected = R"([{"subject":"1","values":[1]}, {"subject":"2","values":[2]}, {"subject":"3","values":[3]}])";
+    auto generate_values = [] () -> coroutine::experimental::generator<int> {
+        for (int i : {1,2,3}) {
+            co_yield i;
+        }
+    };
+    class generate_json_values {
+        coroutine::experimental::generator<int> _gen;
+        std::optional<int> _next;
+    public:
+        generate_json_values(coroutine::experimental::generator<int> gen) : _gen(std::move(gen)) {}
+        future<bool> has_value() {
+            _next = co_await _gen();
+            co_return _next.has_value();
+        }
+        future<> write(output_stream<char>& out) {
+            object_json obj;
+            obj.subject = std::to_string(*_next);
+            obj.values.push(*_next);
+            co_await json::formatter::write(out, obj);
+        }
+    };
+    formatter_check_expected(expected, [&] (auto& out) {
+        auto gen = generate_json_values(generate_values());
+        generate_json_array(out, gen).get();
+    }, true);
+}
+
 SEASTAR_THREAD_TEST_CASE(formatter_write) {
 
     formatter_check_expected("3", [] (auto &out) {
```

avikivity commented:
Right, for small objects it shouldn't switch at all, and for large object one switch per task-quota.

bhalevy commented:
In v2 ([b5119db](https://github.com/scylladb/seastar/commit/b5119db598d7202491fda779c30a4b511420f6b7)):
- Improved doc comments
- Added support for buffered generator

bhalevy commented:
In v3 ([1ea1719](https://github.com/scylladb/seastar/commit/1ea1719ae14a78b19b1315ee05ca4bb18ca02975)):
- call `generate_values_buffered<circular_buffer>` with explicit template parameter

bhalevy commented:
[f6971c0](https://github.com/scylladb/seastar/commit/f6971c0b9a4286f648f172d6ea88a43a90467ea0):
- rebased
- dropped default template parameter for the `generate_values_buffered` template (clang++ didn't like it)

avikivity commented on include/seastar/json/json_elements.hh:
```diff
@@ -367,6 +377,47 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
+ * onto the \c out output_stream as a json array.
+ */
+template<Jsonable T>
+future<> generate_array_element(output_stream<char>& out, coroutine::experimental::generator<T>& gen) {
```
I think we should accept it by value since we consume it.

bhalevy replied:
Also, the generator not consumed by generate_array_element.
The latter is called in an async loop by the function returned by `generate_array`:
```c++
    return [&gen] (output_stream<char>&& s) mutable -> future<> {
        auto out = std::move(s);
        auto f = co_await coroutine::as_future(generate_array_element(out, gen));
        co_await out.close();
        f.get();
    };
```

avikivity replied:
Can we return noncopyable_function?

Maybe precluded by another requirement, too bad.

avikivity replied:
```c++
struct json_return_type {
    sstring _res;
    std::function<future<>(output_stream<char>&&)> _body_writer;
    json_return_type(std::function<future<>(output_stream<char>&&)>&& body_writer) : _body_writer(std::move(body_writer)) {
    }
```

Let's do this:
1. Add a type alias for `std::function<future<> (output_stream<char>&&)>`
2. Add a new API_LEVEL where this type alias is a noncopyable_function

avikivity replied:
To reduce future churn, place this in the internal subnamespace.

bhalevy replied:
@avikivity do you think the api change should be part of this series or as a follow up?

Here, the returned function captures the generator by reference and the generator is kept alive by the caller so it's ok to return the coroutine lambda as a copyable function.

avikivity replied:
It could be a follow up.

avikivity replied:
However, do place internal functions in an internal subnamespace.

avikivity commented on include/seastar/json/json_elements.hh:
```diff
@@ -367,6 +377,47 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
+ * So the caller is responsible for ensuring that the generator remains valid for the lifetime of the returned function.
+ */
+template<Jsonable T>
+std::function<future<>(output_stream<char>&&)> generate_array(coroutine::experimental::generator<T>& gen) {
```
Again, by value.

bhalevy commented on include/seastar/json/json_elements.hh:
```diff
@@ -232,6 +236,12 @@ public:
 
+template <typename T>
+concept Jsonable = requires (T j) {
+    { j.to_json() } -> std::convertible_to<std::string>;
```
done in [c8448d8](https://github.com/scylladb/seastar/commit/c8448d845ac12e522a9f51710fb09f7dca5127d2)

bhalevy commented on include/seastar/json/json_elements.hh:
```diff
@@ -232,6 +236,12 @@ public:
+template <typename T>
+concept Jsonable = requires (T j) {
+    { j.to_json() } -> std::convertible_to<std::string>;
+    { j.write(std::declval<output_stream<char>&>()) } -> std::same_as<future<>>;
```
done in [c8448d8](https://github.com/scylladb/seastar/commit/c8448d845ac12e522a9f51710fb09f7dca5127d2)

avikivity commented:
> It looks like the build issue with clang happens with clang version 18.1.3 but is fixed in later version (19.1.7) as in scylla dbuild environment or 20.1.7 on Fedora 42.

I'll regenerate the toolchain to include clangs 19 and 20.

bhalevy commented:
@avikivity please re-review and consider merging

avikivity commented on include/seastar/json/json_elements.hh:
```diff
@@ -367,6 +377,47 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
 }
 
+/*!
+ * \brief consume values from a coroutine generator \c gen and encode the items using the \c fun mapping function
```
There's no fun any more.

avikivity commented on include/seastar/json/json_elements.hh:
```diff
@@ -367,6 +377,47 @@ std::function<future<>(output_stream<char>&&)> stream_range_as_array(Container v
+ */
+template<Jsonable T>
+std::function<future<>(output_stream<char>&&)> generate_array(coroutine::experimental::generator<T>& gen) {
+    return [&gen] (output_stream<char>&& s) mutable -> future<> {
```
Don't we need coroutine::lambda here?

bhalevy replied:
Good point

bhalevy replied:
Indeed there is a general issue with copying std::function holding a lambda coroutine that I reproduced, but having the std::function hold a coroutine::lambda doesn't help either.
The coroutine::lambda is required when generating the continuation task that would invoke the std::function eventually.

At any rate, it seems like it doesn't make a difference here since gen is captured by reference in the lambda function.

For completeness, here is a bad example with stack-use-after-return:
```
    double n = dist(rand_eng);
    auto make_fun = [n] () -> std::function<future<double>()> {
        return [n] () -> future<double> {
            // Expect coroutine capture to be freed after co_await without coroutine::lambda
            co_await yield();
            co_return n;
        };
    };
    auto fun = make_fun();
    auto n2 = co_await yield().then([fun] {
        return fun();
    });
```

Bad example 2:
```
    double n = dist(rand_eng);
    auto make_fun = [n] () -> std::function<future<double>()> {
        return coroutine::lambda([n] () -> future<double> {
            // Expect coroutine capture to be freed after co_await without coroutine::lambda
            co_await yield();
            co_return n;
        });
    };
    auto fun = make_fun();
    auto n2 = co_await yield().then([fun] {
        return fun();
    });
```

Good example 1:
```
    auto make_fun = [n] () -> std::function<future<double>()> {
        return [n] () -> future<double> {
            // Expect coroutine capture to be freed after co_await without coroutine::lambda
            co_await yield();
            co_return n;
        };
    };
    auto fun = make_fun();
    auto n2 = co_await yield().then([&fun] {
        return fun();
    });
```

Good example 2:
```
    double n = dist(rand_eng);
    auto make_fun = [n] () -> std::function<future<double>()> {
        return [n] () -> future<double> {
            // Expect coroutine capture to be freed after co_await without coroutine::lambda
            co_await yield();
            co_return n;
        };
    };
    auto fun = make_fun();
    auto n2 = co_await yield().then(coroutine::lambda([fun] {
        return fun();
    }));
```

bhalevy commented:
In [8dcc2c3](https://github.com/scylladb/seastar/commit/8dcc2c3cf0008fc3e082faff4bd2e02fe2385f65):
- Fixed documentation comments
- Modified added unit tests to test copying of the function returned by `generate_array`

bhalevy commented:
@avikivity ping, please re-review and consider merging

avikivity commented:
I don't understand why the lambda coroutine doesn't explode.

bhalevy commented:
> I don't understand why the lambda coroutine doesn't explode.

I think that since `gen` is captured by reference, it is not lost when the lambda returns on `co_await`, since the compiler refers to the calling function's frame to get to `gen` rather than keeping a variable with `&gen`'s value in the lambda function captures.

bhalevy commented:
In [63dc17f](https://github.com/scylladb/seastar/commit/63dc17f5d4bc2bbffd7cc79aa9ee6ef04945630b):
- Refactored patches for better clarity and bisectibility
- util: noncopyable_function: enable_lw_shared_from_this
  - to help make `http::shared_body_writer_type = lw_shared_ptr<noncopyable_function<future<>(output_stream<char>&&)>>`
- pass the the coroutine generator by value to `generate_array` and capture it by value now that the returned coroutine lambda function is kept as a shared noncopyable function

avikivity commented on include/seastar/util/noncopyable_function.hh:
```diff
@@ -104,7 +105,7 @@ struct is_nothrow_if_object<> {
 SEASTAR_MODULE_EXPORT
 template <typename Ret, typename... Args, bool Noexcept>
-class noncopyable_function<Ret (Args...) noexcept(Noexcept)> : private internal::noncopyable_function_base {
+class noncopyable_function<Ret (Args...) noexcept(Noexcept)> : private internal::noncopyable_function_base, public enable_lw_shared_from_this<noncopyable_function<Ret (Args...) noexcept(Noexcept)>> {
```
A common low-level utility should not assume anything about its callers.

bhalevy replied:
Dropped in [fa7f581](https://github.com/scylladb/seastar/commit/fa7f5815d8a31e808a3129b15312717071bdf551)

bhalevy commented:
[fa7f581](https://github.com/scylladb/seastar/commit/fa7f5815d8a31e808a3129b15312717071bdf551):
- rebased
- dropped "util: noncopyable_function: enable_lw_shared_from_this"

avikivity commented on include/seastar/http/reply.hh:
```diff
@@ -282,7 +286,7 @@ private:
     future<> write_reply_headers(httpd::connection& connection);
 
-    noncopyable_function<future<>(output_stream<char>&&)> _body_writer;
+    http::shared_body_writer_type _body_writer;
```
Should be the nonshared type (or the API-selected type).

avikivity commented on CMakeLists.txt:
```diff
@@ -70,10 +70,10 @@ option (Seastar_DEPRECATED_OSTREAM_FORMATTERS
 
 set (Seastar_API_LEVEL
-  "7"
+  "8"
```
> json: API level 8: use noncopyable_function in json_return_type
> Hold a shared_body_writer_type so we can copy json_return_type
> without copying the body_writer function.
> 
> This will allow returning a co_routine lambda as body_writer
> that can be safely held as a lw_shard_ptr<noncopyabe_function>
> in the following patch.
> 

Why does json_return_type need to be copyable?

bhalevy commented on include/seastar/http/reply.hh:
```diff
@@ -245,7 +245,11 @@ struct reply {
-    void write_body(const sstring& content_type, noncopyable_function<future<>(output_stream<char>&&)>&& body_writer);
+    void write_body(const sstring& content_type, http::shared_body_writer_type body_writer);
+
+    void write_body(const sstring& content_type, http::body_writer_type&& body_writer) {
```
One is a shared ptr the other is passed as rvalue reference meaning the callee assumes ownership of the object, while in the shared ptr case the writer is held by the caller.

bhalevy commented on include/seastar/json/json_elements.hh:
```diff
@@ -315,10 +316,20 @@ struct json_void : public jsonable{
     json_return_type(body_writer_type&& body_writer) : _body_writer(std::move(body_writer)) {
     }
+#endif
+
```
We did open a can of worms...
I'll take a stab at it and push a simplified api with the new api level that will allow the callee to hold a noncopyable body writer function and let request and reply use it by reference.
My main worry there is lifecycle management to ensure the function is kept alive while requests and replies using it may use it.

nyh replied:
> json_return_type encapsulate an object that can be as small as an int, and here we are with gates and shared pointers to noncopyable functions and confusion.
> 
> Better to start from scratch. This is an API, the users should be able to use it without having to decide so many things.

By the way, this is what I ended up doing in some Scylla Alternator code that was using json_return_type, and I realized the existing json_return_type is simply not good enough - I wanted it to use a non-copyable function, and couldn't. So I discovered that although Seastar supports json_return_type in some things, I didn't actually need to use it for anything, everything works without it and in fact everything became simpler by not using it :-) See https://github.com/scylladb/scylladb/pull/24480/commits/d8fab2a01a459b01e88b16c8d096268074707166.

avikivity replied:
That's also a route we should consider.

bhalevy replied:
Simplified version pushed in [fa00799](https://github.com/scylladb/seastar/commit/fa00799b8ddabe6ebb28bbffa200154ae6c3a980)

bhalevy commented:
In [fa00799](https://github.com/scylladb/seastar/commit/fa00799b8ddabe6ebb28bbffa200154ae6c3a980):
- rebased
- simplified model where only `body_writer_type = noncopyable_function<future<>(output_stream<char>&&)>` is used
  - With that json_return_type became noncopyable
  - Disabling `httpd_test.test_shared_future`
  - Adjusted `httpd_test.json_test_handler` to call `rep->write_body` with a lambda calling the noncopyable `http::body_writer_type _write_func` member, instead of implicitly generating a noncopyable_function from a std::function member

bhalevy commented:
> > In [fa00799](https://github.com/scylladb/seastar/commit/fa00799b8ddabe6ebb28bbffa200154ae6c3a980):
> > 
> > * rebased
> > * simplified model where only `body_writer_type = noncopyable_function<future<>(output_stream<char>&&)>` is used
> >   
> >   * With that json_return_type became noncopyable
> >   * Disabling `httpd_test.test_shared_future`
> >   * Adjusted `httpd_test.json_test_handler` to call `rep->write_body` with a lambda calling the noncopyable `http::body_writer_type _write_func` member, instead of implicitly generating a noncopyable_function from a std::function member
> 
> @bhalevy let's split off the json_return_type changes, they're a lot more involved than generate_array.

@avikivity made this PR a draft until https://github.com/scylladb/seastar/pull/2921 that focuses on the json_return_type changes.

avikivity commented on tests/unit/json_formatter_test.cc:
```diff
@@ -135,6 +141,30 @@ SEASTAR_THREAD_TEST_CASE(test_stream_range_as_array) {
+    formatter_check_expected(expected, [streamer = std::move(streamer)] (auto& out) {
+        streamer(std::move(out)).get();
+    }, false);
+}
```
If it's an internal function, it shouldn't be used.

I'm looking for an example that generates something like this:

```json
{
   "a": 4,
   "b": 7,
   "c": [some long array represented by a generator],
   "d": 3
}
```

bhalevy replied:
> If it's an internal function, it shouldn't be used.

So maybe it shouldn't be internal then.

> I'm looking for an example that generates something like this:
> 
> ```json
> {
>    "a": 4,
>    "b": 7,
>    "c": [some long array represented by a generator],
>    "d": 3
> }
> ```

I'm not sure how to do it using the existing api since the returned `body_writer_type` consumes the output stream and closes it.  It is meant to serialize a singular json result, not an element within a larger json object.

`generate_array_element` that you asked to be in the `seastar::internal` namespace does that since it outputs a json array onto the output stream and it doesn't close the stream, so it can be called to compose one json array element as part of a higher level json object.

bhalevy replied:
For scylla API this is sufficient.

One of the elaborate use cases that I'm aware of is describe_ring:
```c++
static future<json::json_return_type> describe_ring_as_json(sharded<service::storage_service>& ss, sstring keyspace) {
    co_return json::json_return_type(stream_range_as_array(co_await ss.local().describe_ring(keyspace), token_range_endpoints_to_json));
}
```
where `describe_ring` returns `utils::chunked_vector<storage_service::token_range_endpoints>`
where:
```c++
struct token_range_endpoints {
    sstring _start_token;
    sstring _end_token;
    std::vector<sstring> _endpoints;
    std::vector<sstring> _rpc_endpoints;
    std::vector<endpoint_details> _endpoint_details;
};
```
and `token_range_endpoints_to_json` returns the interesting json object:
```c++
struct token_range : public json::json_base {
    json::json_element< sstring > start_token;
    json::json_element< sstring > end_token;
    json::json_list< sstring > endpoints;
    json::json_list< sstring > rpc_endpoints;
    json::json_list< endpoint_detail > endpoint_details;
```
using:
```c++
static ss::token_range token_range_endpoints_to_json(const dht::token_range_endpoints& d) {
    ss::token_range r;
    r.start_token = d._start_token;
    r.end_token = d._end_token;
    r.endpoints = d._endpoints;
    r.rpc_endpoints = d._rpc_endpoints;
    for (auto det : d._endpoint_details) {
        ss::endpoint_detail ed;
        ed.host = fmt::to_string(det._host);
        ed.datacenter = det._datacenter;
        if (det._rack != "") {
            ed.rack = det._rack;
        }
        r.endpoint_details.push(ed);
    }
    return r;
}
```

Note that `r.endpoint_details` is constructed dynamically, but it is fully materialized.
Usually it has a short enough list so it doesn't matter much, but if it to generate a list of undetermined, and potentially long size, it could use a way for nested streaming of data.

Not sure about Alternator use case.
@nyh can you please elaborate?

avikivity replied:
Can't we have a json object that contains the generated array as one of its elements?

Now that I think of it, why does it have to be a generator? It can be a range, and since std::generator is a range, it just works.

bhalevy replied:
> Can't we have a json object that contains the generated array as one of its elements?
> 
> Now that I think of it, why does it have to be a generator? It can be a range, and since std::generator is a range, it just works.

This change is about an async henerator which isn'r a range unfortunately (it was nicecto have something like that)

bhalevy replied:
@avikivity can we proceed with this PR as it is?

avikivity replied:
It seems like a very partial solution.

It feels like we're piling on bandaids instead of making a working whole.
